sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/windowing/AfterProcessingTime.java

```diff
@@ -17,40 +17,87 @@
  */
 package org.apache.beam.sdk.transforms.windowing;
 
+import com.google.common.collect.ImmutableList;
+import java.util.Collections;
 import java.util.List;
+import java.util.Locale;
 import java.util.Objects;
 import org.apache.beam.sdk.annotations.Experimental;
-import org.apache.beam.sdk.transforms.SerializableFunction;
-import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.transforms.windowing.Trigger.OnceTrigger;
+import org.joda.time.Duration;
 import org.joda.time.Instant;
+import org.joda.time.format.PeriodFormat;
+import org.joda.time.format.PeriodFormatter;
 
 /**
  * {@code AfterProcessingTime} triggers fire based on the current processing time. They operate in
  * the real-time domain.
- *
- * <p>The time at which to fire the timer can be adjusted via the methods in
- * {@link AfterDelayFromFirstElement}, such as {@link AfterDelayFromFirstElement#plusDelayOf} or
- * {@link AfterDelayFromFirstElement#alignedTo}.
  */
 @Experimental(Experimental.Kind.TRIGGER)
-public class AfterProcessingTime extends AfterDelayFromFirstElement {
+public class AfterProcessingTime extends OnceTrigger {
+
+  private static final PeriodFormatter PERIOD_FORMATTER = PeriodFormat.wordBased(Locale.ENGLISH);
+
+  private final List<TimestampTransform> timestampTransforms;
 
-  private AfterProcessingTime(List<SerializableFunction<Instant, Instant>> transforms) {
-    super(TimeDomain.PROCESSING_TIME, transforms);
+  public AfterProcessingTime(List<TimestampTransform> timestampTransforms) {
+    super(null);
+    this.timestampTransforms = timestampTransforms;
   }
 
   /**
-   * Creates a trigger that fires when the current processing time passes the processing time
-   * at which this trigger saw the first element in a pane.
+   * Creates a trigger that fires when the current processing time passes the processing time at
+   * which this trigger saw the first element in a pane.
    */
   public static AfterProcessingTime pastFirstElementInPane() {
-    return new AfterProcessingTime(IDENTITY);
+    return new AfterProcessingTime(Collections.<TimestampTransform>emptyList());
+  }
+
+  /**
+   * The transforms applied to the arrival time of an element to determine when this trigger allows
+   * output.
+   */
+  public List<TimestampTransform> getTimestampTransforms() {
+    return timestampTransforms;
+  }
+
+  /**
+   * Adds some delay to the original target time.
+   *
+   * @param delay the delay to add
+   * @return An updated time trigger that will wait the additional time before firing.
+   */
+  public AfterProcessingTime plusDelayOf(final Duration delay) {
+    return new AfterProcessingTime(
+        ImmutableList.<TimestampTransform>builder()
+            .addAll(timestampTransforms)
+            .add(TimestampTransform.delay(delay))
+            .build());
+  }
+
+  /**
+   * Aligns timestamps to the smallest multiple of {@code size} since the {@code offset} greater
```

Comment:
Yes, `new Instant(0)` is Java for "the timestamp since the epoch" but I will clarify since my wording was really bad.